### ISSUE TITLE
Ideas on a proper allocator

### ISSUE THREAD

**Issue by Z-snails** [Kindelia/HVM#126]
See https://github.com/Kindelia/HVM/issues/38#issuecomment-1109431355

A proper allocator would be nice.
Currently we only have 32 bits to play with in each `Lnk`, this is not enough to use the system allocator on 64 bit platforms.
So either we can
1 - write our own bump/pool allocator
2 - change the memory layout

--------------------

I prefer option 2. With option 2 we could allow up to unboxed `u32` s (on 64 bit machines)
It allows us to use any allocator we want.
It might increase memory usage fairly significantly on 64 bit platforms (~ 50%)

This may also be slightly faster as it reduces memory reads/pointer indirections
My suggestion for the memory layout is the following:

top level
| tagged pointer |

| 32 bits | usize bits ... |
| tag |  arguments ... |

`(StrCons 1 (StrCons 2 (StrCons 3 StrNil)))`
On 32 bit platforms:
currently has size 8 * 7 = 56 bytes
will have 4 * 13 = 52 bytes (less!)

On 64 bit platforms:
currently has size 8 * 7 = 56 bytes
will have size 8 * 11 = 88 bytes

------

Another option for the memory layout
On 64 bit platforms have 96 bit pointers.

| 32 bit tag | 64 bit pointer |

| 32 bit tag | 32 bits tag | 64 bit pointer | 64 bit pointer |
| arg 1 tag | arg 2 tag | arg 1 | arg 2 |

This way we can combine the tag for arg 1 and arg 2 into a single u64, and keep aligned loads/stores.
This also allows unboxed `u64` s.

`(StrCons 1 (StrCons 2 (StrCons 3 StrNil)))`
Would with this layout have size: 12 + 3 * 8 * 3 = 84 bytes

**Comment by VictorTaelin** [comment]
The HVM was updated to 60 bits and, while ugly-looking, is here to stay. It is basically as much as we can get while still keeping the 4-bit tag. Not sure I get your new memory layout suggestion though. Can you rephrase?

**Comment by Z-snails** [comment]
Well with the 60 bit pointers, that is enough to use the system allocator, so long as allocations are 16 byte aligned.

This means the changes I suggested aren't useful anymore.

C11 specifies the aligned_alloc function, which is available from glibc 2.16 or there is posix_memalign which is part of POSIX.

**Comment by VictorTaelin** [comment]
I've tried many alternative allocators and the silly one we have seems to perform the best. I'm open to changes, not sure how though.
